File: pkg/cloudfilename/cloudfilename_test.go
// Copyright 2026 The Cocomhub Authors. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package cloudfilename

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// TestDefaultFromURL_FromFixture 以共享语料 testdata/cases.json 为基准验证 DefaultFromURL。
// 该语料同时被 Web UI 的 JS 单测 (web/static/cloudfilename.test.js) 复用，
// 保证 Go 服务端与浏览器端对同一 URL 推导出完全一致的默认文件名。
func TestDefaultFromURL_FromFixture(t *testing.T) {
	cases := loadFixture(t)
	for url, want := range cases {
		if got := DefaultFromURL(url); got != want {
			t.Errorf("DefaultFromURL(%q) = %q, want %q", url, got, want)
		}
	}
}

// TestDefaultFromURL_KeyRules 对关键 wget 规则做显式断言，便于读测试即理解行为。
func TestDefaultFromURL_KeyRules(t *testing.T) {
	tests := []struct {
		name string
		url  string
		want string
	}{
		{name: "普通文件", url: "https://example.com/file.txt", want: "file.txt"},
		{name: "多级目录取最后一段", url: "https://example.com/a/b/c.jpg", want: "c.jpg"},
		{name: "根路径", url: "https://example.com/", want: "index.html"},
		{name: "根路径带查询", url: "https://example.com/?a=v", want: "index.html_a=v"},
		{name: "目录结尾使用index.html", url: "https://example.com/foo/", want: "index.html"},
		{name: "目录结尾带查询", url: "https://example.com/xx/?a=v", want: "index.html_a=v"},
		{name: "查询参数直接附加", url: "https://example.com/file.txt?token=abc&x=1", want: "file.txt_token=abc&x=1"},
		{name: "百分号解码", url: "https://example.com/my%20file.txt", want: "my file.txt"},
		{name: "百分号解码不转加号", url: "https://example.com/a+b.txt", want: "a+b.txt"},
		{name: "双重编码", url: "https://example.com/a%2520b.txt", want: "a b.txt"},
		{name: "无路径", url: "https://example.com", want: "index.html"},
		{name: "无路径带查询", url: "https://example.com?a=v", want: "index.html_a=v"},
		{name: "非法百分号编码回退download", url: "https://example.com/100%.txt", want: "download"},
		{name: "中文路径解码", url: "https://example.com/%E4%B8%AD%E6%96%87.txt", want: "中文.txt"},
		{name: "无效URL", url: "not a url", want: "download"},
		{name: "相对URL无host", url: "example.com/file.txt", want: "download"},
		{name: "空串", url: "", want: "download"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := DefaultFromURL(tt.url); got != tt.want {
				t.Errorf("DefaultFromURL(%q) = %q, want %q", tt.url, got, tt.want)
			}
		})
	}
}

func TestResolveFilename_ExplicitValid(t *testing.T) {
	got, err := ResolveFilename(Entry{URL: "https://e.com/a.zip", Filename: "valid.zip"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != "valid.zip" {
		t.Fatalf("want valid.zip, got %q", got)
	}
}

func TestResolveFilename_ExplicitUnsafe(t *testing.T) {
	_, err := ResolveFilename(Entry{URL: "https://e.com/a.zip", Filename: "a/b.zip"})
	if err == nil {
		t.Fatal("expected error for unsafe filename")
	}
	if !errors.Is(err, ErrEntryUnsafeFilename) {
		t.Fatalf("want ErrEntryUnsafeFilename, got %v", err)
	}
}

func TestResolveFilename_AutoFromURL(t *testing.T) {
	got, err := ResolveFilename(Entry{URL: "https://e.com/xx/?a=v"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	// DefaultFromURL 现在是安全版，? 被替换为 _
	if got != "index.html_a=v" {
		t.Fatalf("want index.html_a=v, got %q", got)
	}
}

func TestValidateEntry(t *testing.T) {
	tests := []struct {
		name string
		e    Entry
		err  error
	}{
		{name: "空 URL", e: Entry{URL: ""}, err: ErrEntryEmptyURL},
		{name: "非法 scheme", e: Entry{URL: "ftp://e.com/a.zip"}, err: ErrEntryBadScheme},
		{name: "缺 host", e: Entry{URL: "http:///path"}, err: ErrEntryMissingHost},
		{name: "合法 URL", e: Entry{URL: "https://e.com/a.zip"}, err: nil},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ValidateEntry(tt.e)
			if tt.err == nil && got != nil {
				t.Fatalf("unexpected error: %v", got)
			}
			if tt.err != nil && !errors.Is(got, tt.err) {
				t.Fatalf("want %v, got %v", tt.err, got)
			}
		})
	}
}

func TestValidateEntries_DupURL_DiffFilename(t *testing.T) {
	entries := []Entry{
		{URL: "https://e.com/a.zip", Filename: "a.zip"},
		{URL: "https://e.com/a.zip", Filename: "b.zip"},
	}
	err := ValidateEntries(entries)
	if !errors.Is(err, ErrEntryDupURL) {
		t.Fatalf("want ErrEntryDupURL, got %v", err)
	}
}

func TestValidateEntries_DupURL_SameFilenameOK(t *testing.T) {
	entries := []Entry{
		{URL: "https://e.com/a.zip", Filename: "a.zip"},
		{URL: "https://e.com/a.zip", Filename: "a.zip"},
	}
	if err := ValidateEntries(entries); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestValidateEntries_Valid(t *testing.T) {
	entries := []Entry{
		{URL: "https://e.com/a.zip"},
		{URL: "https://e.com/b.zip", Filename: "b.zip"},
	}
	if err := ValidateEntries(entries); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestSafe(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want string
	}{
		{name: "普通文件名不变", in: "file.txt", want: "file.txt"},
		{name: "反斜杠替换", in: `a\b.txt`, want: "a_b.txt"},
		{name: "斜杠替换", in: "a/b.txt", want: "a_b.txt"},
		{name: "问号替换", in: "a?b.txt", want: "a_b.txt"},
		{name: "冒号替换", in: "a:b.txt", want: "a_b.txt"},
		{name: "尖括号替换", in: "a<b>c", want: "a_b_c"},
		{name: "竖线替换", in: "a|b", want: "a_b"},
		{name: "双引号替换", in: `a"b`, want: "a_b"},
		{name: "星号替换", in: "a*b", want: "a_b"},
		{name: "Tab替换", in: "a	b", want: "a_b"},
		{name: "NUL移除", in: "a\x00b", want: "ab"},
		{name: "组合替换", in: `a/b?c:d"e*f|g<h>i`, want: "a_b_c_d_e_f_g_h_i"},
		{name: "等号与与号保留", in: "file.txt?a=b&c=d", want: "file.txt_a=b&c=d"},
		{name: "首尾空白去除", in: "  file.txt  ", want: "file.txt"},
		{name: "首尾点去除", in: "..file.txt..", want: "file.txt"},
		{name: "仅点与空白", in: "... ...", want: "download"},
		{name: "空串", in: "", want: "download"},
		{name: "Windows保留名CON", in: "CON", want: "_CON"},
		{name: "Windows保留名小写带扩展", in: "con.txt", want: "_con.txt"},
		{name: "Windows保留名COM1", in: "COM1", want: "_COM1"},
		{name: "Windows保留名LPT9", in: "lpt9", want: "_lpt9"},
		{name: "CON后接字符不命中", in: "CONtext.txt", want: "CONtext.txt"},
		{name: "超长文件名截断保留扩展名", in: strings.Repeat("a", 300) + ".txt", want: strings.Repeat("a", 250) + ".txt"},
		{name: "超长多字节不劈开UTF-8", in: strings.Repeat("好", 100) + ".zip", want: strings.Repeat("好", 83) + ".zip"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := Safe(tt.in); got != tt.want {
				t.Errorf("Safe(%q) = %q, want %q", tt.in, got, tt.want)
			}
		})
	}
}

func TestDefaultFromURL_SafeOutput(t *testing.T) {
	tests := []struct{ url, want string }{
		{"https://e.com/path/file.txt?x=1&y=2", "file.txt_x=1&y=2"},
		{"https://e.com/a?b/c", "a_b_c"},
	}
	for _, tt := range tests {
		if got := DefaultFromURL(tt.url); got != tt.want {
			t.Errorf("DefaultFromURL(%q) = %q, want %q", tt.url, got, tt.want)
		}
	}
}

func TestDefaultFromURL_UnsafeRaw(t *testing.T) {
	// defaultFromURLUnsafe 保留原始 wget 语义：? 不会被替换
	got := defaultFromURLUnsafe("https://e.com/xx/?a=v")
	if want := "index.html?a=v"; got != want {
		t.Errorf("defaultFromURLUnsafe = %q, want %q", got, want)
	}
}

// TestDefaultFromURLThenSafe 验证"生成 + 清理"的完整链路（与 server 端一致）。
// DefaultFromURL 已内置 Safe，双重包装退化为单层 Safe，结果不变。
func TestDefaultFromURLThenSafe(t *testing.T) {
	tests := []struct {
		url  string
		want string
	}{
		{url: "https://example.com/xx/?a=v", want: "index.html_a=v"},
		{url: "https://example.com/a%2Fb.txt", want: "b.txt"},
		{url: "https://example.com/my%20file.txt", want: "my file.txt"},
		{url: "https://example.com/path/file.txt?x=1", want: "file.txt_x=1"},
	}
	for _, tt := range tests {
		if got := Safe(DefaultFromURL(tt.url)); got != tt.want {
			t.Errorf("Safe(DefaultFromURL(%q)) = %q, want %q", tt.url, got, tt.want)
		}
	}
}

// loadFixture 读取共享语料 testdata/cases.json。
func loadFixture(t *testing.T) map[string]string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("testdata", "cases.json"))
	if err != nil {
		t.Fatalf("读取语料失败: %v", err)
	}
	var cases map[string]string
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("解析语料失败: %v", err)
	}
	return cases
}

File: web/static/cloudfilename.test.js
/* SPDX-License-Identifier: Apache-2.0 */
/*
 * cloudfilename.test.js —— 浏览器端文件名生成/清理规则的单元测试。
 *
 * 运行方式（仓库根目录）：
 *   node --test web/static/cloudfilename.test.js
 *
 * 语料与 Go 端共享：pkg/cloudfilename/testdata/cases.json。
 * Go 测试 pkg/cloudfilename 与本测试都断言该语料，从而保证双端对同一
 * URL 推导出完全一致的默认文件名（wget 行为）。
 */
'use strict';

const test = require('node:test');
const assert = require('node:assert');
const path = require('node:path');
const fs = require('node:fs');

const { genDefaultFilename, filepathSafe, safeDefaultFromURL, validateEntry, validateEntries } = require('./cloudfilename.js');

const fixturePath = path.join(__dirname, '../../pkg/cloudfilename/testdata/cases.json');
const cases = JSON.parse(fs.readFileSync(fixturePath, 'utf8'));

test('safeDefaultFromURL 与 Go pkg/cloudfilename 共享语料一致', () => {
  for (const [url, want] of Object.entries(cases)) {
    assert.strictEqual(safeDefaultFromURL(url), want, `URL: ${url}`);
  }
});

test('filepathSafe 清理规则与 Go pkg/cloudfilename.Safe 一致', () => {
  const safeCases = [
    ['a/b.txt', 'a_b.txt'],
    ['a?b', 'a_b'],
    ['a:b', 'a_b'],
    ['..file.txt..', 'file.txt'],
    ['... ...', 'download'],
    ['', 'download'],
    ['index.html_a=v', 'index.html_a=v'],
    ['my file.txt', 'my file.txt'],
    ['a' + String.fromCharCode(0) + 'b', 'ab'],
    ['  file.txt  ', 'file.txt'],
    // Windows 保留设备名
    ['CON', '_CON'],
    ['con.txt', '_con.txt'],
    ['COM1', '_COM1'],
    ['lpt9', '_lpt9'],
    ['CONtext.txt', 'CONtext.txt'], // CON 后接其他字符不命中
    // 超长文件名按字节截断（保留扩展名，不劈开 UTF-8）
    ['a'.repeat(300) + '.txt', 'a'.repeat(250) + '.txt'],
    ['好'.repeat(100) + '.zip', '好'.repeat(83) + '.zip'],
  ];
  for (const [input, want] of safeCases) {
    assert.strictEqual(filepathSafe(input), want, `input: ${JSON.stringify(input)}`);
  }
});

test('genDefaultFilename + filepathSafe 完整链路', () => {
  const chainCases = [
    ['https://example.com/xx/?a=v', 'index.html_a=v'],
    ['https://example.com/a%2Fb.txt', 'b.txt'],
    ['https://example.com/my%20file.txt', 'my file.txt'],
    ['https://example.com/path/file.txt?x=1', 'file.txt_x=1'],
  ];
  for (const [url, want] of chainCases) {
    assert.strictEqual(filepathSafe(genDefaultFilename(url)), want, `URL: ${url}`);
  }
});

test('safeDefaultFromURL 安全语义（? 被替换为 _）', () => {
  assert.strictEqual(safeDefaultFromURL('https://example.com/xx/?a=v'), 'index.html_a=v');
  assert.strictEqual(safeDefaultFromURL('https://example.com/file.txt?x=1'), 'file.txt_x=1');
});

test('validateEntry URL 格式校验（结构化结果）', () => {
  assert.deepStrictEqual(validateEntry(''), { valid: false, code: 'EMPTY_URL', message: 'URL is empty' });
  assert.deepStrictEqual(validateEntry('ftp://e.com/a.zip'), { valid: false, code: 'BAD_SCHEME', message: 'unsupported URL scheme (only http/https)' });
  assert.deepStrictEqual(validateEntry('https://e.com/a.zip'), { valid: true, code: 'OK', message: '' });
  // 尾随空白并入路径，不影响校验（Go url.Parse 同样把尾随空白当作路径 → 合法）。
  assert.deepStrictEqual(validateEntry('https://e.com/a.zip '), { valid: true, code: 'OK', message: '' });
  // 前导空白：JS 的正则要求 scheme 位于开头 → BAD_SCHEME；
  // Go url.Parse 报错（first path segment…colon）→ 同样拒绝。两者 reject 语义一致。
  assert.strictEqual(validateEntry(' http://e.com/a.zip').valid, false);
});

test('validateEntries 同 URL 不同 filename 去重', () => {
  assert.strictEqual(validateEntries([{ url: 'https://e.com/a.zip' }]).valid, true);
  assert.strictEqual(
    validateEntries([
      { url: 'https://e.com/a.zip', filename: 'a.zip' },
      { url: 'https://e.com/a.zip', filename: 'a.zip' },
    ]).valid, true);
  const dup = validateEntries([
    { url: 'https://e.com/a.zip', filename: 'a.zip' },
    { url: 'https://e.com/a.zip', filename: 'b.zip' },
  ]);
  assert.strictEqual(dup.valid, false);
  assert.strictEqual(dup.code, 'DUP_URL');
  // 单条目 URL 非法时返回该条目的错误
  const bad = validateEntries([{ url: '' }]);
  assert.strictEqual(bad.valid, false);
  assert.strictEqual(bad.code, 'EMPTY_URL');
});
